torch/csrc/jit/codegen/cuda/evaluator_common.cpp

```diff
@@ -0,0 +1,481 @@
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+template <typename VALTYPE>
+std::vector<VALTYPE*> getImmediateProducers(VALTYPE* val) {
+  if (val->definition()) {
+    auto expr = val->definition();
+    return expr->inputs();
+  } else {
+    return {};
+  }
+}
+
+//! IR-Generic utility, collects all the producers required for the
+//!  given list of IR values and returns them along with the original
+//!  list in topological order.
+template <typename VALTYPE>
+std::vector<VALTYPE*> makeSortedEvaluationList(std::vector<VALTYPE*> input) {
+  // Deduplicate
+  std::vector<VALTYPE*> to_sort;
+  std::unordered_set<VALTYPE*> visited;
+  for (auto val : input) {
+    if (!visited.count(val)) {
+      to_sort.push_back(val);
+      visited.insert(val);
+    }
+  }
+
+  std::vector<VALTYPE*> sorted;
+  visited.clear();
+
+  // Topological Sort
+  //  Note: didn't explicitly exclude producers that are not in the original
+  //   list. This should be acceptable for the intended use.
+  while (!to_sort.empty()) {
+    auto top_val = to_sort.back();
+    if (visited.count(top_val)) {
+      to_sort.pop_back();
+    } else {
+      bool ready_to_pop = true;
+      for (auto producer : getImmediateProducers(top_val)) {
+        if (!visited.count(producer)) {
+          ready_to_pop = false;
+          to_sort.push_back(producer);
+        }
+      }
+      if (ready_to_pop) {
+        visited.insert(top_val);
+        sorted.push_back(top_val);
+        to_sort.pop_back();
+      }
+    }
+  }
+
+  return sorted;
+}
+
+//! Kernel IR utility, collects all the symbolic integers
+//!  used in allocation nodes.
+void collectBufferSizes(
+    std::vector<kir::Val*>& into,
+    const std::vector<kir::Expr*>& exprs) {
+  for (auto expr : exprs) {
+    if (auto allocate = dynamic_cast<kir::Allocate*>(expr)) {
+      into.push_back(allocate->size());
+    } else if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      collectBufferSizes(into, for_loop->body().exprs());
+    } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+      collectBufferSizes(into, ite->thenBody().exprs());
+      collectBufferSizes(into, ite->elseBody().exprs());
+    }
+  }
+}
+
+//! Kernel IR utility, collects all the kir symbolic
+//!  integers we will need at runtime, i.e. after the
+//!  generated cuda kernel has already been compiled.
+//!  The values are to be used for runtime logic, like
+//!  `computeLaunchparams`.
+std::vector<kir::Val*> collectRuntimeUsedIntegers(
+    Fusion* fusion,
+    GpuLower* lower) {
+  std::vector<kir::Val*> ret;
+
+  // Collect extent and integer inputs
+  for (auto val : fusion->usedMathVals()) {
+    auto kir_val = lower->lowerValue(val);
+    if (auto kir_tv = dynamic_cast<kir::TensorView*>(kir_val)) {
+      for (auto id : kir_tv->domain()->domain()) {
+        ret.push_back(id->extent());
```

Comment:
Can we filter this by ids bound to a thread dimension?

Reply:
I think so at this point since we only use parallelized extents at runtime currently. This part is meant to capture all symbolic integers that'd be needed on the runtime path so it could be updated often.

Reply:
I have added filtering to include just about enough integers for runtime usage and the list could get longer for future constructs.

Reply:
Actually it does look risky to apply this filtering since we also use this path to evaluate buffer sizes and for dynamic smem allocation we seem to allow serial IterDomains to be allocated. 

I guess we could keep all extents on the workspace in this PR and could try to minimize it later if needed. It didn't seem to be becoming a major bottleneck on runtime path short term.

Reply:
Thank you, I think that makes sense. I'm sure we could try to filter them better, but I assume we'd have to somehow look at what is actually used. Makes sense to leave it for now.